In [15]:
import pandas as pd 

data  = pd.read_csv('/mnt/c/Users/dell/Downloads/Tech-Projects-Portfolio/Data-Internship-Home-Assignment/source/jobs.csv')
data.columns = ['index', 'context']
data.head(10)

index                                            context
0      0  {"@context": "http://schema.org", "@type": "Jo...
1      1  {"@context": "http://schema.org", "@type": "Jo...
2      2  {"@context": "http://schema.org", "@type": "Jo...
3      3  {"@context": "http://schema.org", "@type": "Jo...
4      4  {"@context": "http://schema.org", "@type": "Jo...
5      5  {"@context": "http://schema.org", "@type": "Jo...
6      6  {"@context": "http://schema.org", "@type": "Jo...
7      7  {"@context": "http://schema.org", "@type": "Jo...
8      8  {"@context": "http://schema.org", "@type": "Jo...
9      9  {"@context": "http://schema.org", "@type": "Jo...

In [5]:
data['context']

0       {"@context": "http://schema.org", "@type": "Jo...
1       {"@context": "http://schema.org", "@type": "Jo...
2       {"@context": "http://schema.org", "@type": "Jo...
3       {"@context": "http://schema.org", "@type": "Jo...
4       {"@context": "http://schema.org", "@type": "Jo...
                              ...                        
8256    {"@context": "http://schema.org", "@type": "Jo...
8257    {"@context": "http://schema.org", "@type": "Jo...
8258    {"@context": "http://schema.org", "@type": "Jo...
8259                                                  NaN
8260    {"@context": "http://schema.org", "@type": "Jo...
Name: context, Length: 8261, dtype: object

### Checking duplicates & Missing Values

In [16]:
data.duplicated().sum()

0

In [17]:
data.isnull().sum()

index        0
context    576
dtype: int64

In [18]:
data.dropna(axis=0, inplace=True)

In [19]:
data.isnull().sum()

index      0
context    0
dtype: int64

In [11]:
data

index                                            context
0         0  {"@context": "http://schema.org", "@type": "Jo...
1         1  {"@context": "http://schema.org", "@type": "Jo...
2         2  {"@context": "http://schema.org", "@type": "Jo...
3         3  {"@context": "http://schema.org", "@type": "Jo...
4         4  {"@context": "http://schema.org", "@type": "Jo...
...     ...                                                ...
8255   8255  {"@context": "http://schema.org", "@type": "Jo...
8256   8256  {"@context": "http://schema.org", "@type": "Jo...
8257   8257  {"@context": "http://schema.org", "@type": "Jo...
8258   8258  {"@context": "http://schema.org", "@type": "Jo...
8260   8260  {"@context": "http://schema.org", "@type": "Jo...

[7685 rows x 2 columns]

In [26]:
import os

for i, row in data.iterrows():
    with open(os.path.join('/mnt/c/Users/dell/Downloads/Tech-Projects-Portfolio/Data-Internship-Home-Assignment/my/extract', f'{i}.txt'), 'w') as f:
        f.write(row['context'])
    f.close()

Let's take a sample and explore how the json file is structured 


```json
{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-18T18:11:34.000Z',
 'description': '<strong><u>Company Description<br><br></u></strong>As the world&apos;s leader in digital payments technology, Visa&apos;s mission is to connect the world through the most creative, reliable and secure payment network - enabling individuals, businesses, and economies to thrive. Our advanced global processing network, VisaNet, provides secure and reliable payments around the world, and is capable of handling more than 65,000 transaction messages a second. The company&apos;s dedication to innovation drives the rapid growth of connected commerce on any device, and fuels the dream of a cashless future for everyone, everywhere. As the world moves from analog to digital, Visa is applying our brand, products, people, network and scale to reshape the future of commerce.<br><br>At Visa, your individuality fits right in. Working here gives you an opportunity to impact the world, invest in your career growth, and be part of an inclusive and diverse workplace. We are a global team of disruptors, trailblazers, innovators and risk-takers who are helping drive economic growth in even the most remote parts of the world, creatively moving the industry forward, and doing meaningful work that brings financial literacy and digital commerce to millions of unbanked and underserved consumers.<br><br><strong>You&apos;re</strong> an Individual. <strong>We&apos;re</strong> the team for you. <strong>Together</strong>, let&apos;s transform the way the world pays.<br><br><strong><u>Job Description<br><br></u></strong>The fraud and risk management team within CyberSource is responsible for designing and building machine learning models for detecting, preventing and managing fraud in e-commerce payment events. The team collaborates closely with Visa colleagues and other internal stakeholders to provide the advanced and multilayered fraud and risk management solutions to enterprise as well as small-to-midsized businesses.<br><br>We are currently seeking a Data Scientist to support fraud risk modeling projects and drive insights for CyberSource. Ideal candidates for this position will have a very strong background of statistics and machine learning techniques and tools and. In addition, the candidate should have a proactive mindset for innovative researches and a highly collaborative approach to project work.<br><br><strong><u>You Will Work Collaboratively With a Team Composed Of Both Data Scientists And Engineers To<br><br></u></strong>The position will be based at Visa’s office in Foster City, California.<br><ul><li>Build and validate risk models with advanced machine learning techniques</li><li>Support model installations, and monitor and calibrate production models</li><li>Interpret and present modeling and analytical results to all level audience</li><li>Define financial and analytic metrics to measure development and production outcomes and produce performance reports</li><li>Conduct transaction data analyses for internal and external product owners, and develop deeper insights into the products using advanced statistical methods</li><li>Support sales and marketing efforts with sound statistical and financial analysis; execute ad-hoc analyses to meet the fast-changing market demands</li><li>Drive analytic product development via conducting statistical analyses on various data sources; and add values to products by innovatively applying the analysis</li><li>Enhance existing solutions and approaches by promoting new methodology and best practices in data science field</li><li>Collaborate with internal and external stakeholders and cross-functional teams</li></ul><strong>Qualifications<br><br></strong><strong><u>Basic Qualifications<br></u></strong><ul><li>2 years of work experience with a Bachelor’s Degree or an Advanced Degree (e.g. Masters, MBA, JD, MD, or PhD)<br></li></ul><strong><u>Preferred Qualifications<br></u></strong><ul><li>Graduate/Post Graduate degree (Master&apos;s or Ph.D.) in quantitative field such as statistics, mathematics, computer science, finance, economics or relevant area.</li><li>Hands-on experience in data science and analytical functions</li><li>Ability to analyze large datasets, extract information, features, anomalies, etc.</li><li>Ability to understand customer/client behavior, apply modeling and machine learning to business problems</li><li>Ability to communicate model outputs, performance internally and externally</li><li>Self-starter, curious, team player in dynamic environments.</li></ul><strong>Additional Information<br><br></strong><strong><u>Travel Requirements<br><br></u></strong>This position requires the incumbent to travel for work up to 5% of the time.<br><br><strong><u>Mental/Physical Requirements<br><br></u></strong>This position will be performed in an office setting. The position will require the incumbent to sit and stand at a desk, communicate in person and by telephone, frequently operate standard office equipment, such as telephones and computers, reach with hands and arms, and bend or lift up to 25 pounds.<br><br><strong><u>EEO Statement<br><br></u></strong>Visa will consider for employment qualified applicants with criminal histories in a manner consistent with EEOC guidelines and applicable local law.<br><br>Job Number: REF35768X',
 'employmentType': 'FULL_TIME',
 'hiringOrganization': {'@type': 'Organization',
  'name': 'Visa',
  'sameAs': 'https://www.linkedin.com/company/visa',
  'logo': 'https://media-exp3.licdn.com/dms/image/C4D0BAQHW9Vyrl6r5ZA/company-logo_200_200/0/1519856212168?e=1634774400&amp;v=beta&amp;t=ByTenFtCKnSoj6tvr8aAp6UFa-hWDycmYddiu7dfPiw'},
 'identifier': {'@type': 'PropertyValue',
  'name': 'Visa',
  'value': 'REF35768X'},
 'image': 'https://media-exp3.licdn.com/dms/image/C4D0BAQHW9Vyrl6r5ZA/company-logo_100_100/0/1519856212168?e=1634774400&v=beta&t=Ef1tp_ZNDeHj2z1otKy6ewIikIoqp2C-QGKGvTP0qO0',
 'industry': 'Consumer Services,Financial Services,Information Technology and Services',
 'jobLocation': {'@type': 'Place',
  'address': {'@type': 'PostalAddress',
   'addressCountry': 'US',
   'addressLocality': 'Foster City',
   'addressRegion': 'CA',
   'postalCode': '94029',
   'streetAddress': None},
  'latitude': 37.561882,
  'longitude': -122.26644},
 'skills': '',
 'title': 'Data Scientist',
 'validThrough': '2021-08-19T03:17:26.000Z',
 'educationRequirements': {'@type': 'EducationalOccupationalCredential',
  'credentialCategory': 'bachelor degree'},
 'experienceRequirements': {'@type': 'OccupationalExperienceRequirements',
  'monthsOfExperience': 24},
 'estimatedSalary': {'@type': 'MonetaryAmount',
  'currency': 'USD',
  'value': {'@type': 'QuantitativeValue',
   'minValue': 101000,
   'maxValue': 156000,
   'unitText': 'YEAR'}}}

   ```

In [21]:
template = {
    "job": {
        "title": None,
        "industry": None,
        "description": None,
        "employment_type": None,
        "date_posted": None,
    },
    "company": {
        "name": None,
        "link": None,
    },
    "education": {
        "required_credential": None,
    },
    "experience": {
        "months_of_experience": None,
        "seniority_level": None,
    },
    "salary": {
        "currency": None,
        "min_value": None,
        "max_value": None,
        "unit": None,
    },
    "location": {
        "country": None,
        "locality": None,
        "region": None,
        "postal_code": None,
        "street_address": None,
        "latitude": None,
        "longitude": None,
    },
}

In [22]:
import html

def text_decoder(text):
    decoded_text = html.unescape(text)
    utf8_text = decoded_text.encode('utf-8')

    return utf8_text.decode('utf-8')

In [28]:
import json 

data_as_json = []
for file in os.listdir(os.path.join('/mnt/c/Users/dell/Downloads/Tech-Projects-Portfolio/Data-Internship-Home-Assignment/my/extract')):
    with open(os.path.join('/mnt/c/Users/dell/Downloads/Tech-Projects-Portfolio/Data-Internship-Home-Assignment/my/extract', file), 'r') as f:
        # Load the JSON data from the file
        data = json.load(f)
        for key in data.keys():
            if isinstance(data[key], str):
                data[key] = text_decoder(data[key])
        data_as_json.append(data)
    f.close()

In [31]:
data_as_json[800]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-31T07:01:00.000Z',
 'description': "If you are a react/C# developer looking for a small, close knit team where you can work remoote, please read on!<br><br>Top Reasons to Work with Us<br><ul><li> Competitive pay, top-tier benefits, and unlimited PTO</li><li> People first environment with great office perks</li><li> 401k Matching!<br></li></ul>What You Will Be Doing<br><br>Full-stack software development<br>Working closely with the engineering team, QA, and product users on a highly functional agile team<br>Leading and guiding a team of junior and mid-level developers and test engineers<br>Ensuring system health and fixing defects<br>Solving difficult problems through research and innovation<br><br>What You Need for this Position<br><br><strong><u>At Least 3 Years Of Experience And Knowledge Of<br></u></strong><ul><li> .NET</li><li> C#</li><li> JavaScript</li><li> REACT</li><li> Azure</li><li> Cloud<br></l

In [37]:
data_as_json[1]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-09T17:39:11.000Z',
 'description': "Ethicon, part of Johnson &amp; Johnson Medical Devices Companies, is recruiting for a <strong>Principal Software Engineer</strong>, located <strong><em><u>remotely</u></em> in the</strong> <strong>United States </strong>or in the<strong> Munich, Germany </strong>area.<br><br>Johnson &amp; Johnson Medical Devices is focused on shaping the future of digital surgery and expanding its robotics and digital solutions offerings across the entire portfolio, with multi-specialty, end-to-end solutions in orthopedics, endoluminal intervention and general surgery. This includes the MONARCH® platform, a first-of-its-kind robotic technology indicated for bronchoscopic visualization of and access to patient airways for diagnostic and therapeutic procedures in the lung.<br><br>Ethicon, part of Johnson &amp; Johnson Medical Devices Companies, has made significant contributions to surger

In [32]:
len(data_as_json)

7685

In [35]:
data_as_json[7684]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-18T18:11:34.000Z',
 'description': '<strong><u>Company Description<br><br></u></strong>As the world&apos;s leader in digital payments technology, Visa&apos;s mission is to connect the world through the most creative, reliable and secure payment network - enabling individuals, businesses, and economies to thrive. Our advanced global processing network, VisaNet, provides secure and reliable payments around the world, and is capable of handling more than 65,000 transaction messages a second. The company&apos;s dedication to innovation drives the rapid growth of connected commerce on any device, and fuels the dream of a cashless future for everyone, everywhere. As the world moves from analog to digital, Visa is applying our brand, products, people, network and scale to reshape the future of commerce.<br><br>At Visa, your individuality fits right in. Working here gives you an opportunity to impact the world, i

In [36]:
8260 - 7684

576

the missing values deleted ealrlier in the data is 576 plus

In [42]:
job_mapping = {
    'date_posted': 'datePosted',
    'description': 'description',
    'employment_type': 'employmentType',
    'industry': 'industry',
    'title': 'title'
}
location_mapping = {
    'country': 'jobLocation.address.addressCountry',
    'latitude': 'jobLocation.latitude',
    'locality': 'jobLocation.address.addressLocality',
    'longitude': 'jobLocation.longitude',
    'postal_code': 'jobLocation.address.postalCode',
    'region': 'jobLocation.address.addressRegion',
    'street_address': 'jobLocation.address.streetAddress',
}
company_mapping = {
    'link': 'hiringOrganization.sameAs',
    'name': 'hiringOrganization.name',
}
salary_mapping = {
    'currency': 'estimatedSalary.currency',
    'max_value': 'estimatedSalary.value.maxValue',
    'min_value': 'estimatedSalary.value.minValue',
    'unit': 'estimatedSalary.value.unitText',
}

In [43]:
for key_in_template, key_in_data in job_mapping.items():
        if key_in_data in data_as_json[0].keys():
            template['job'][key_in_template] = data_as_json[0][key_in_data]

In [44]:
template

{'job': {'title': 'Senior Data Engineer',
  'industry': 'Broadcast Media',
  'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design and development of data pip

In [52]:
for key_in_template, key_in_data in company_mapping.items():

    print(key_in_template ,key_in_data )
        

link hiringOrganization.sameAs
name hiringOrganization.name


In [60]:
data_as_json[7684].keys()

dict_keys(['@context', '@type', 'datePosted', 'description', 'employmentType', 'hiringOrganization', 'identifier', 'image', 'industry', 'jobLocation', 'skills', 'title', 'validThrough', 'educationRequirements', 'experienceRequirements', 'estimatedSalary'])

In [64]:
if "hiringOrganization" in data_as_json[7684].keys():
            print("sdkljb")

sdkljb


In [59]:
for key_in_template, key_in_data in company_mapping.items():
        print(key_in_data)
        keys = key_in_data.split('.')
        print(keys)
        value = data_as_json[7684]
        if key_in_data in data_as_json[7684].keys():
            print(key_in_data)
            
            print(keys)
            for key in keys:
                if key in value:
                    value = value[key]
                else:
                    break
            else:
                print("dslkjbf")
                template['company'][key_in_template] = value
                template

hiringOrganization.sameAs
['hiringOrganization', 'sameAs']
hiringOrganization.name
['hiringOrganization', 'name']


In [46]:
template

{'job': {'title': 'Senior Data Engineer',
  'industry': 'Broadcast Media',
  'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design and development of data pip

In [47]:
data_as_json[0]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-09T17:39:28.000Z',
 'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design an

In [66]:
len(template)

6

In [71]:
def extract_nested_value(data, splitted_keys):
    for key in splitted_keys:
        if key in data:
            data = data[key]
        else:
            return None
    return data

job_mapping = {
    'date_posted': 'datePosted',
    'description': 'description',
    'employment_type': 'employmentType',
    'industry': 'industry',
    'title': 'title'
}
location_mapping = {
    'country': 'jobLocation.address.addressCountry',
    'latitude': 'jobLocation.latitude',
    'locality': 'jobLocation.address.addressLocality',
    'longitude': 'jobLocation.longitude',
    'postal_code': 'jobLocation.address.postalCode',
    'region': 'jobLocation.address.addressRegion',
    'street_address': 'jobLocation.address.streetAddress',
}
company_mapping = {
    'link': 'hiringOrganization.sameAs',
    'name': 'hiringOrganization.name',
}
salary_mapping = {
    'currency': 'estimatedSalary.currency',
    'max_value': 'estimatedSalary.value.maxValue',
    'min_value': 'estimatedSalary.value.minValue',
    'unit': 'estimatedSalary.value.unitText',
}
transformed_data = []
for data in data_as_json:
    fill_template = template
    for key_in_template, key_in_data in job_mapping.items():
        fill_template['job'][key_in_template] = data.get(key_in_data, None)

    for key_in_template, key_in_data in company_mapping.items():
        fill_template['company'][key_in_template] = extract_nested_value(data, key_in_data.split('.'))

    fill_template['education']['required_credential'] = extract_nested_value(data, 'educationRequirements.credentialCategory'.split('.'))

    if 'experienceRequirements' in data and isinstance(data['experienceRequirements'], dict):
        fill_template['experience']['months_of_experience'] = data['experienceRequirements'].get('monthsOfExperience', None)
    fill_template['experience']['seniority_level'] = "NA"

    for key_in_template, key_in_data in location_mapping.items():
        fill_template['location'][key_in_template] = extract_nested_value(data, key_in_data.split('.'))

    for key_in_template, key_in_data in salary_mapping.items():
        fill_template['salary'][key_in_template] = extract_nested_value(data, key_in_data.split('.'))
    
    transformed_data.append(template)

In [72]:
transformed_data[0]

{'job': {'title': 'Data Scientist',
  'industry': 'Consumer Services,Financial Services,Information Technology and Services',
  'description': '<strong><u>Company Description<br><br></u></strong>As the world&apos;s leader in digital payments technology, Visa&apos;s mission is to connect the world through the most creative, reliable and secure payment network - enabling individuals, businesses, and economies to thrive. Our advanced global processing network, VisaNet, provides secure and reliable payments around the world, and is capable of handling more than 65,000 transaction messages a second. The company&apos;s dedication to innovation drives the rapid growth of connected commerce on any device, and fuels the dream of a cashless future for everyone, everywhere. As the world moves from analog to digital, Visa is applying our brand, products, people, network and scale to reshape the future of commerce.<br><br>At Visa, your individuality fits right in. Working here gives you an opportun

In [73]:
template

{'job': {'title': 'Data Scientist',
  'industry': 'Consumer Services,Financial Services,Information Technology and Services',
  'description': '<strong><u>Company Description<br><br></u></strong>As the world&apos;s leader in digital payments technology, Visa&apos;s mission is to connect the world through the most creative, reliable and secure payment network - enabling individuals, businesses, and economies to thrive. Our advanced global processing network, VisaNet, provides secure and reliable payments around the world, and is capable of handling more than 65,000 transaction messages a second. The company&apos;s dedication to innovation drives the rapid growth of connected commerce on any device, and fuels the dream of a cashless future for everyone, everywhere. As the world moves from analog to digital, Visa is applying our brand, products, people, network and scale to reshape the future of commerce.<br><br>At Visa, your individuality fits right in. Working here gives you an opportun

In [74]:
mapping = {
    'job': {
        'date_posted': 'datePosted',
        'description': 'description',
        'employment_type': 'employmentType',
        'industry': 'industry',
        'title': 'title',
    },
    'location': {
        'country': 'jobLocation.address.addressCountry',
        'latitude': 'jobLocation.latitude',
        'locality': 'jobLocation.address.addressLocality',
        'longitude': 'jobLocation.longitude',
        'postal_code': 'jobLocation.address.postalCode',
        'region': 'jobLocation.address.addressRegion',
        'street_address': 'jobLocation.address.streetAddress',
    },
    'company': {
        'link': 'hiringOrganization.sameAs',
        'name': 'hiringOrganization.name',
    },
    'salary': {
        'currency': 'estimatedSalary.currency',
        'max_value': 'estimatedSalary.value.maxValue',
        'min_value': 'estimatedSalary.value.minValue',
        'unit': 'estimatedSalary.value.unitText',
    },
    'education': {
        'required_credential': 'educationRequirements.credentialCategory',
    },
    'experience': {
        'months_of_experience': 'experienceRequirements.monthsOfExperience',
        'seniority_level': "I don't know",
    },
}
transformed_data = []
for data in data_as_json:
    template = {
        "job": {
            "title": None,
            "industry": None,
            "description": None,
            "employment_type": None,
            "date_posted": None,
        },
        "company": {
            "name": None,
            "link": None,
        },
        "education": {
            "required_credential": None,
        },
        "experience": {
            "months_of_experience": None,
            "seniority_level": None,
        },
        "salary": {
            "currency": None,
            "min_value": None,
            "max_value": None,
            "unit": None,
        },
        "location": {
            "country": None,
            "locality": None,
            "region": None,
            "postal_code": None,
            "street_address": None,
            "latitude": None,
            "longitude": None,
        },
    }
    for section, section_mapping in mapping.items():
        for key_in_template, key_in_data in section_mapping.items():
            keys = key_in_data.split('.')
            value = data
            for key in keys:
                if key in value:
                    value = value[key]
                else:
                    break
            else:
                template[section][key_in_template] = value
    transformed_data.append(template)


In [76]:
transformed_data[0]

{'job': {'title': 'Senior Data Engineer',
  'industry': 'Broadcast Media',
  'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design and development of data pip

In [77]:
len(data_as_json)

7685

In [78]:
len(transformed_data)

7685

In [79]:
data_as_json[0]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-09T17:39:28.000Z',
 'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design an

In [80]:
"""<p><strong><u>ABOUT US:</u></strong></p><p>As the leader of rehabilitative care, Encompass Health offers both facility-based and home-based patient care through its 
national network of rehabilitation hospitals, home health agencies, and hospice agencies that spans 42 states and Puerto Rico. Setting the standard for providing excellent 
care, Encompass Health has earned its place among Modern Healthcare’s “Best Places to Work in Healthcare” and the Fortune “World’s Most Admired Companies” for 2021.
</p><p><br></p><p><strong><u>POSITION PURPOSE:</u></strong></p><p>The&nbsp;<strong>HRIS Security &amp; Data Maintenance Analyst</strong>&nbsp;plays an integral role in HR 
Systems security definition and is responsible for defining, setting up, maintaining and accurately documenting security access for all HR systems. The HRIS Security &amp; 
Data Maintenance Analyst also oversees application support including table setup and maintenance, data integrity, and auditing of data. Additionally, this role will assist 
with the overall administrative support of existing HR systems including upgrades, conversions, and problem resolution.</p><p><br></p><p><strong><u>RESPONSIBILITES &amp; 
TASKS:</u></strong></p><p>SECURITY &amp; AUDITS:</p><ul><li>Defines and sets up security roles and permissions for HR Systems.</li><li>Defines and maintains dynamic role 
and permission assignment for auto provisioning and de-provisioning.</li><li>Creates and maintains a crosswalk that outlines role permissions and assigned titles.</li><li>
Creates queries/reports that extract role definitions and permissions for auditing purposes.</li><li>Acts as security administrator by processing access requests to HR 
applications and systems.</li><li>Ensures appropriate approvals are received for any ad hoc requests.</li><li>Serves as the SME that partners with various business areas to 
provide guidance on best practices and current HR security protocols.</li><li>Performs periodic audits of HR system user security/access.</li><li>Validates current company 
structure against employee job movement to maintain continuous access accuracy.</li><li>Looks for opportunities to increase efficiencies for common requests.</li><li>Supports 
employees/end users with HR system access issues.</li><li>Adheres to department’s SLA for processing requests.</li></ul><p><br></p><p>TABLE MAINTENANCE, UPLOADS &amp; DATA 
INTEGRITY AUDITS:</p><ul><li>Responsible for mass data uploads, coordinating with HR contacts to ensure accuracy in the process.</li><li>Performs data structure and integrity 
audits on employee data mass upload requests.</li><li>Seeks to minimize errors and determine impacts to downstream systems/processes, as well as impacts to employees due to 
requested changes/movement.</li><li>Updates system fields and tables associated with business changes and operations.</li><li>Maintains tables and security changes within HR 
Systems.</li><li>Proactively utilizes PeopleSoft Query, OBI Reporting or SQL tools to identify data integrity issues.</li><li>Determines the appropriate approach to ensure 
employee and company structure data elements are updated correctly.</li><li>Conducts data audits to identify potential non-compliance. I.e. personal data audits…duplicate 
NID, DOB out of guidelines.</li><li>Partners with internal audit to research and analyze potential data discrepancies.</li><li>Partners with ITG to implement automated access 
audit processes.</li><li>Participates in new projects through testing/implementation as it relates to access and security.</li><li>Completes mandatory training and courses 
required by completion date.</li><li>Reports questionable situations, concerns, complaints or harassment immediately.</li><li>Conducts audits of system process completions 
and reports out to business owners. i.e. Onboarding not complete.</li><li>Administers duplicate processing in ATS and other HR systems.</li><li>Ensures compliance regarding 
HR information sharing, management, organization, hierarchy determination, retrieval, accessibility, dissemination, sharing, and quality control.
</li></ul><p><br></p><p><strong><u>Qualifications:</u></strong></p><p>Minimum Qualifications:</p><ul><li>Bachelor's degree preferred. Associate's degree with 4 years of 
related experience may be substituted for Bachelor's degree.</li><li>Four or more years of experience required in related field (i.e. Human Resources Operations, HRIS, HR 
Technology, IT, etc.).</li><li>Strong knowledge of Microsoft Word, Excel, and PowerPoint. Advanced Excel skills preferred. Strong analytic and diagnostic skills as well as 
a data-driven and metrics-driven approach.</li><li>Minimum 3 years of experience with PeopleSoft version 9.2 or higher preferred.</li><li>Experience with workflow analysis, 
business process analysis, finding improvement areas, and use of business modeling tools.</li><li>Experience with security and auditing processes.</li></ul><p><br></p><p>
Machines, Equipment Used:</p><ul><li>General office equipment such as computer/laptop, telephone, copy/fax machine, calculator, scanner, etc.</li><li>Microsoft Office 
software, to include Outlook, Word, PowerPoint, and Excel.</li></ul><p><br></p><p>Physical Requirements:</p><ul><li>Visual acuity, speech recognition, speech clarity.
</li><li>Ability to lift, lower, push, pull, and retrieve objects weighing a minimum of 15 pounds. Reasonable assistance may be requested when lifting, pushing, and/or 
pulling exceeds these minimum requirements.</li></ul><p><br></p><p>Skills and Abilities:</p><ul><li>Oral communication, written communication, fluency in English, active 
listening.</li><li>Information ordering, deductive reasoning, social perceptiveness, time management, critical thinking.</li><li>Ability to coordinate, analyze, observe, 
make decisions, and meet deadlines in a detail-oriented manner.</li><li>Ability to work independently without continuous supervision.</li><li>Ability to understand and 
react quickly to changing priorities.</li><li>Demonstrated accuracy and thoroughness while meeting customer needs in a timely manner.</li></ul><p><br></p><p>Environmental 
Conditions:</p><ul><li>Indoor, temperature controlled, smoke-free environment.</li><li>Handicapped accessible.</li><li>May work under stressful circumstances at times.
</li></ul><p><br></p><p>Proficiency or Productivity Standards:</p><ul><li>Has regular, reliable, and predictable attendance and punctuality.</li><li>Adheres to dress code 
including wearing ID badge.</li><li>Adheres to Standards of Business Conduct.</li><li>May be required to work weekdays and/or weekends, evenings and/or night shifts if needed 
to meet deadlines.</li><li>May be required to work on religious and/or legal holidays on scheduled days/shifts.</li><li>May be required to perform other duties as assigned 
by supervisor.</li><li>This position will support cultural diversity by promoting and maintaining an inclusive work environment and culture that is respectful and accepting 
of diversity.</li></ul><p></p>",
 """

'<p><strong><u>ABOUT US:</u></strong></p><p>As the leader of rehabilitative care, Encompass Health offers both facility-based and home-based patient care through its \nnational network of rehabilitation hospitals, home health agencies, and hospice agencies that spans 42 states and Puerto Rico. Setting the standard for providing excellent \ncare, Encompass Health has earned its place among Modern Healthcare’s “Best Places to Work in Healthcare” and the Fortune “World’s Most Admired Companies” for 2021.\n</p><p><br></p><p><strong><u>POSITION PURPOSE:</u></strong></p><p>The&nbsp;<strong>HRIS Security &amp; Data Maintenance Analyst</strong>&nbsp;plays an integral role in HR \nSystems security definition and is responsible for defining, setting up, maintaining and accurately documenting security access for all HR systems. The HRIS Security &amp; \nData Maintenance Analyst also oversees application support including table setup and maintenance, data integrity, and auditing of data. Additiona

In [81]:
"""Tinder brings people together. With tens of millions of users and a presence in every country on earth, our reach is expansive and rapidly growing.
<br><br>The Analytics team powers data-informed decision-making at Tinder. We are looking to add a hardworking, creative Data Scientist to our growing team.
As a Data Scientist on Analytics, you will partner with multiple teams - including Product, Marketing, Engineering, and Finance - to improve our extraordinary platform.
You will use this comprehensive dataset to develop experiments, forecasting models, automated reports, and various exploratory insights to guide our business decisions.
<br><br>An effective Data Scientist at Tinder must have a passion and aptitude for turning sophisticated problems into concrete hypotheses that can be answered through 
thorough data analysis and A/B testing. They must also be able to disseminate their takeaways and recommendations to a variety of partners within our organization via 
presentations, reports, and dashboards. In this role, your insights will influence both the performance of our business and the experience of our users.<br><br><strong>
Responsibilities:<br></strong><li>Be a leader with a point of view for the organization, using data to constantly challenge and reshape our current understanding of 
customer behavior and business performance</li><li>Derive and communicate data-driven recommendations regarding both the prioritization of potential product/marketing
initiatives and performance of current initiatives</li><li>Define and operationalize the detailed tracking of company-wide, team-specific, and product-specific 
performance metrics via rollup tables, dashboards, and automated reporting</li><li>Design and analyze experiments to rigorously test a specific hypothesis</li><li>
Find the “story” in the data and readily share insights across the organization<br><br></li><strong>Skills Required:<br></strong><li>No formal industry experience is 
required - we are looking for creative, tenacious problem solvers who are passionate about our product and excited to work with our rich, complex dataset</li><li>Ability
to develop a nuanced business understanding, testable hypotheses, and meaningful findings from data</li><li>Eagerness to identify and explore new product/marketing 
opportunities, potential improvements to our data infrastructure, and unanswered questions about customer behavior</li><li>Outstanding written and verbal communication
skills, and a willingness to collaborate with business partners</li><li>Comfort using descriptive and inferential statistics</li><li>Experience with SQL and data 
visualization tools (e.g. Tableau or Mode) is highly recommended</li><li>Familiarity with scripting languages, machine learning, advanced statistical techniques, 
and A/B testing is a plus<br><br></li><strong>As part of our team, you’ll enjoy:<br></strong><li> Working on a product that has an immediate impact on people’s 
lives all around the world</li><li> Collaborating with a team of creative, fun and driven colleagues</li><li> Comprehensive health coverage, competitive salary,
401(k) employer match, Employee Stock Purchase Plan (ESPP)</li><li> Other perks and wellness benefits like a fitness membership subsidy, paid concierge medical 
membership, pet insurance offerings, and a commuter subsidy</li><li> A robust Learning + Development offering including our MentorMatch program, access to a 
library of 6,000+ online learning courses through Udemy, and an annual L+D stipend</li><li> Access to mental health resources</li><li> Fertility preservation 
benefits</li><li> No Meeting Wednesdays and access to a wide range of product and service discounts through Perkspot</li><li> Charitable donations match up 
to $15,000 annually</li><li> Monthly and weekly interactive virtual events including Book Club, trivia with prizes and yoga workouts</li><li> The opportunity
to join six active Employee Resource Groups (ERGs)<br><br></li>At Tinder, we don’t just accept difference — we celebrate it, we support it, and we thrive 
on it for the benefit of our employees, our product, and our community. We strive to make our workplace an inclusive and diverse environment, giving people 
from all walks of life the opportunity to have a voice. We champion and encourage those who bring different perspectives, ideas, and creativity to join our 
team dedicated to bringing people together across the globe. Tinder is proud to be an equal opportunity workplace where we welcome all people regardless of 
sex, gender identity, race, ethnicity, disability, or other lived experience.',
 """

"Tinder brings people together. With tens of millions of users and a presence in every country on earth, our reach is expansive and rapidly growing.\n<br><br>The Analytics team powers data-informed decision-making at Tinder. We are looking to add a hardworking, creative Data Scientist to our growing team.\nAs a Data Scientist on Analytics, you will partner with multiple teams - including Product, Marketing, Engineering, and Finance - to improve our extraordinary platform.\nYou will use this comprehensive dataset to develop experiments, forecasting models, automated reports, and various exploratory insights to guide our business decisions.\n<br><br>An effective Data Scientist at Tinder must have a passion and aptitude for turning sophisticated problems into concrete hypotheses that can be answered through \nthorough data analysis and A/B testing. They must also be able to disseminate their takeaways and recommendations to a variety of partners within our organization via \npresentations

In [82]:
data_as_json[15]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-02T15:14:42.000Z',
 'description': '<p>Hi,</p><p>&nbsp;</p><p>Hope you are doing well,&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</p><p>&nbsp;</p><p>Our client is hiring for <strong>Data Engineer</strong> candidate for <strong>Seattle, WA or Burbank/Glendale/Anaheim, CA (Remote Start)</strong> location<strong>. </strong>Please review the below job description and share your updated resume, if you are available and interested in a long term contract role.</p><p><strong>&nbsp;</strong></p><p><strong>Position:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Data Engineer</strong></p><p><strong>Location:&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Seattle, WA or Burbank/Glendale/Anaheim CA (Remote Start)</strong></p><p><strong>Duration:&nbsp;&nbsp;&

In [83]:
data_as_json[200]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-21T01:46:47.000Z',
 'description': '<strong><u>Summary<br><br></u></strong>Provides and/or facilitates the planning, analysis, design, development, implementation, and maintenance of software applications.<br><br><strong>DUTIES<br></strong><ul><li>Participate in the backlog identification and refinement for new and/or existing systems.</li><li>Work with other Application Development team members to estimate, plan and schedule tasks to meet assigned target dates. </li><li>Document new and existing systems, develop workflows, and create other supporting documents in adherence to all departmental and corporate standards.</li><li>Obtain approvals and consult with production control as needed to prepare programs/systems for implementation; correspond with clients to implement operations programs/enhancements and identify/resolve problems.</li><li>Review completed code, documentation and other deliverables to e

In [84]:
data_as_json[300]

{'@context': 'http://schema.org',
 '@type': 'JobPosting',
 'datePosted': '2021-07-19T22:01:55.000Z',
 'description': "<strong><u>Our Company<br><br></u></strong>Changing the world through digital experiences is what Adobe’s all about. We give everyone—from emerging artists to global brands—everything they need to design and deliver exceptional digital experiences! We’re passionate about empowering people to create beautiful and powerful images, videos, and apps, and transform how companies interact with customers across every screen.<br><br>We’re on a mission to hire the very best and are committed to creating exceptional employee experiences where everyone is respected and has access to equal opportunity. We realize that new ideas can come from everywhere in the organization, and we know the next big idea could be yours!<br><br><strong>The Opportunity <br><br></strong>Adobe is looking for a Machine Learning Engineer who will apply AI and machine learning techniques to big-data problem

I think we can extract from the description skills , and we can add it to the template ! 

but How , we can interact with an LLM to do that ! Let's try it and see

In [96]:
transformed_data[0]

{'job': {'title': 'Senior Data Engineer',
  'industry': 'Broadcast Media',
  'description': "Job Title: Senior Data Engineer<br>Location: Alexandria, VA<br>Salary Range: $120k - $150k<br>Requirements: ETL/ELT, SQL, AWS/Google Cloud, Linux/Unix, Spark (preferred), NoSQL (preferred), Machine Learning concepts (preferred)<br><br>Based in beautiful Alexandria, VA, we are one of the hottest media analytics and software start-ups in the DC area. Due to growth, we are actively seeking to hire a Senior Data Engineer to join our team. The ideal candidate will have at least 5 years of experience with data pipelines (built with Python), ETL experience, a strong SQL background, experience with AWS or Google Cloud, and strong Unix/Linux fundamentals. Any experience with Spark, NoSQL, and Machine Learning would be a huge plus. If this sounds like you, please apply now or send your resume to shiv.warrier@cybercoders.com!<br><br>What You Will Be Doing<br><ul><li>Lead design and development of data pip

In [95]:
import csv


# Extracted information
job_info = sample_data['job']
company_info = sample_data['company']
education_info = sample_data['education']
experience_info = sample_data['experience']
salary_info = sample_data['salary']
location_info = sample_data['location']

# Save as JSON
json_data = {
    'job': job_info,
    'company': company_info,
    'education': education_info,
    'experience': experience_info,
    'salary': salary_info,
    'location': location_info
}

with open('output.json', 'w') as json_file:
    json.dump(json_data, json_file, indent=4)

# Save as CSV
csv_data = [job_info, company_info, education_info, experience_info, salary_info, location_info]
csv_keys = max((d.keys() for d in csv_data), key=len)

with open('output.csv', 'w', newline='') as csv_file:
    writer = csv.DictWriter(csv_file, fieldnames=csv_keys)
    writer.writeheader()
    writer.writerows(csv_data)

CSV file '/mnt/c/Users/dell/Downloads/Tech-Projects-Portfolio/Data-Internship-Home-Assignment/my/extract/llm.csv' has been created.


In [85]:
template = {
            "job": {
                "title": "None",
                "industry": "None",
                "description": "None",
                "employment_type": "None",
                "date_posted": "None",
            },
            "company": {
                "name": "None",
                "link": "None",
            },
            "education": {
                "required_credential": "None",
            },
            "experience": {
                "months_of_experience": "None",
                "seniority_level": "None",
            },
            "salary": {
                "currency": "None",
                "min_value": "None",
                "max_value": "None",
                "unit": "None",
            },
            "location": {
                "country": "None",
                "locality": "None",
                "region": "None",
                "postal_code": "None",
                "street_address": "None",
                "latitude": "None",
                "longitude": "None",
            },
        }

In [86]:
columns = template.keys()

In [87]:
for col in columns:
    feilds = list(template[col].keys())
    values = ', '.join([f"'{template[col][feild]}'" for feild in feilds])
    feilds = ', '.join(feilds)

    print(f"INSERT INTO {col} ({feilds}) VALUES ({values});")


INSERT INTO job (title, industry, description, employment_type, date_posted) VALUES ('None', 'None', 'None', 'None', 'None');
INSERT INTO company (name, link) VALUES ('None', 'None');
INSERT INTO education (required_credential) VALUES ('None');
INSERT INTO experience (months_of_experience, seniority_level) VALUES ('None', 'None');
INSERT INTO salary (currency, min_value, max_value, unit) VALUES ('None', 'None', 'None', 'None');
INSERT INTO location (country, locality, region, postal_code, street_address, latitude, longitude) VALUES ('None', 'None', 'None', 'None', 'None', 'None', 'None');
